### PR TITLE
Rewrite the C wrapper

### PR DESCRIPTION
**Pull request by odow**
Part of #267 Currently failing because things are still only half-fixed, but it's enough to get a sense for discussion.

I've learnt a lot about the code. The main issues:

 * No C return codes are checked or thrown
 * Multiple ways to do the same thing by (ab)using multiple dispatch
 * Lots of macros to "simplify" the code

It all looks very familiar, because it is exactly how a lot of the early Cbc/Clp/GLPK/Gurobi/CPLEX wrappers were written. In the long run though, I think it was a bad choice. The MILP wrappers are a lot more maintainable now that we switched to `libxxx.jl` + MOI. And you can still call any C function, it just takes a little more overhead, but it's also much clearer exactly what you are doing and why.

### CODE REVIEW

File: src/C_wrapper.jl [hunk 1]
```diff
@@ -297,536 +162,629 @@ function KN_release_license(lm::LMcontext)
...
+    ptr_lambda::Ptr{Cdouble},
+    user_data::Ptr{Cvoid},
+)::Cint
+    model = unsafe_pointer_to_objref(user_data)::Model
+    nx, nc = Ref{Cint}(0), Ref{Cint}(0)
+    KN_get_number_vars(model, nx)
+    KN_get_number_cons(model, nc)
+    x = unsafe_wrap(Array, ptr_x, nx[])
+    lambda = unsafe_wrap(Array, ptr_lambda, nx[] + nc[])
+    return model.ms_initpt_callback(model, nSolveNumber, x, lambda, model.ms_initpt_user_data)
```
Comment (odow) [inline src/C_wrapper.jl:728]:
```suggestion
    return model.ms_initpt_callback(
        model,
        nSolveNumber,
        x,
        lambda,
        model.ms_initpt_user_data,
    )
```

**Comment by odow** [comment]
I've asked the Artelys folks if they have people using the current low-level API. That would be one blocker moving forward with this. (We could keep the old functions if needed. But only advertise the "proper" way of using the API going forward.)

**Comment by mgabay** [comment]
Hi Oscar,

We can't think of any client who could be using it currently.

Anyway, I've been looking into the discussion in #267 and #268 and the proposal makes a lot of sense! I think it is a very good idea to go on with these changes, especially as 1.0 has not been released yet.

Also, I expect it would make it easier to reflect possible additions to Knitro API in the future. Don't worry, no breaking changes coming in foreseeable future on our side ;-)

Thanks a lot for your work on this!

**Comment by odow** [comment]
> Also, I expect it would make it easier to reflect possible additions to Knitro API in the future

Yes, precisely. This change would make it so there is no friction between changes/updates/new features in the C API and Julia. They would be exactly equivalent.

This is the approach we have taken with Gurobi.jl etc, and it works very nicely.

**Comment by odow** [comment]
Since this is getting rather big, I'm going to merge this for now. I'll make a separate PR that fixes all of the return code checking.

**Pull request merged**